atc/creds/dummy/manager_factory.go

```diff
@@ -49,5 +52,11 @@ func (factory *managerFactory) NewInstance(config interface{}) (creds.Manager, e
 		})
 	}
 
+	// delay should only be used in unit tests.
+	if delay, ok := configMap["delay"]; ok {
```

Comment:
It feels pretty hacky to have to pass a `delay` into the config for unit tests. Thinking about it more, I don't totally understand the reasoning for the modifications to `inPoolManager` - it's to avoid blocking long-blocking calls to `NewSecrets`, but seems to add a fair bit of complexity to support the `Secrets` not yet being ready. Looking at the credential managers, though, Vault is the only one that can block on `NewSecrets`.

So, what if we kept `inPoolManager` as it was, and just moved the "wait for login + timeout" code to `Vault.Get` instead? Is there a reason that won't work?

Reply:
@aoldershaw Thanks for the comment. Handing "login timeout" in Vault is a different direction from my current implementation. Please allow me to explain my idea and logic.

`inPoolManager` was only introduced by me for `var_sources`. When I implemented `var_sources`, I followed a principle that it should not touch logic of any individual credential manager.

For the system level credential manager, because "concourse web" takes time between startup and scheduling a build to run, thus login time in most of cases is not a problem. Only Vault credential manager add a sleep in `NewSecrets`, but that couldn't prove other credential managers to not requiring a login time. I only use Vault, I never tested other credential managers.

If kept `inPoolManager` as it was, it would imply that `NewSecrets` should return immediately without having any blocking operations. I'm afraid it would be hard to make all contributors to clearly aware of that, and thus it could be error prone, especially considering more credential managers may be added in future.

With my implementation of this PR, the complexity is contained in `inPoolManager`, thus it's testable and controllable. Changes in `inPoolManager` can be easily tested by unit tests, but changes in credential managers, say Vault, requires end-to-end tests with a real Vault, thus hard to be fully covered by unit tests. Thus I decided to avoid touching credential manager related code.

`dummy` was added @vito for testing `var_sources`, so I feel ok to add a new configuration option `delay`. Maybe I can delete the comment in line 55 that could be confused.

Putting all together, I feel enhancing `inPoolManager` is relative simpler and safer.

Reply:
I hear you, but to me, the fact that `NewSecrets` doesn't return an `error` is a pretty good indication that it shouldn't have blocking operations, as most blocking operations are error-prone - the fact that all other credential manager implementations follow this pattern is an indication of that. The blocking in `VaultFactory.NewSecrets` is super hacky as a result of not being able to return an `error` - if the login (which takes place elsewhere) "times out" (i.e. takes too long from when `NewSecrets` was called, rather than from when the login was actually initiated), we still return a Vault instance - it just won't work properly.

So IMO the Vault implementation should really just be made non-blocking. I don't think this should be that hard. The simplest implementation could just involve threading the `loggedIn` chan to the `Vault` struct, and in `Vault.findSecret`, waiting for the `loggedIn` chan to receive (with a timeout). Potentially a better design would be to pass in a modified `SecretReader` that handles the `ReAuther`'s behaviour and wraps the regular Vault client.

But either way, I think a slight change to the `Vault` implementation would be preferable.

Reply:
Okey.